In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sqlite3
%matplotlib inline

In [5]:
import os
for dirname, _, filenames in os.walk('/Users/luzi/Desktop/doppEx2/database.sqlite'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

In [7]:
# Load your data and print out a few lines. Perform operations to inspect data
#   types and look for instances of missing or possibly errant data.
#load frist table of (player attributes)
db= sqlite3.connect('/Users/luzi/Desktop/doppEx2/database.sqlite')
df=pd.read_sql('SELECT * FROM player_attributes',db)
df.head()

id  player_fifa_api_id  player_api_id                 date  overall_rating  \
0   1              218353         505942  2016-02-18 00:00:00            67.0   
1   2              218353         505942  2015-11-19 00:00:00            67.0   
2   3              218353         505942  2015-09-21 00:00:00            62.0   
3   4              218353         505942  2015-03-20 00:00:00            61.0   
4   5              218353         505942  2007-02-22 00:00:00            61.0   

   potential preferred_foot attacking_work_rate defensive_work_rate  crossing  \
0       71.0          right              medium              medium      49.0   
1       71.0          right              medium              medium      49.0   
2       66.0          right              medium              medium      49.0   
3       65.0          right              medium              medium      48.0   
4       65.0          right              medium              medium      48.0   

   ...  vision  penalties  marking  standing_tackle  sliding_tackle  \
0  ...    54.0       48.0     65.0             69.0            69.0   
1  ...    54.0       48.0     65.0             69.0            69.0   
2  ...    54.0       48.0     65.0             66.0            69.0   
3  ...    53.0       47.0     62.0             63.0            66.0   
4  ...    53.0       47.0     62.0             63.0            66.0   

   gk_diving  gk_handling  gk_kicking  gk_positioning  gk_reflexes  
0        6.0         11.0        10.0             8.0          8.0  
1        6.0         11.0        10.0             8.0          8.0  
2        6.0         11.0        10.0             8.0          8.0  
3        5.0         10.0         9.0             7.0          7.0  
4        5.0         10.0         9.0             7.0          7.0  

[5 rows x 42 columns]

In [9]:
df_player=pd.read_sql('SELECT * FROM player',db)
df_player.head()

id  player_api_id         player_name  player_fifa_api_id  \
0   1         505942  Aaron Appindangoye              218353   
1   2         155782     Aaron Cresswell              189615   
2   3         162549         Aaron Doran              186170   
3   4          30572       Aaron Galindo              140161   
4   5          23780        Aaron Hughes               17725   

              birthday  height  weight  
0  1992-02-29 00:00:00  182.88     187  
1  1989-12-15 00:00:00  170.18     146  
2  1991-05-13 00:00:00  170.18     163  
3  1982-05-08 00:00:00  182.88     198  
4  1979-11-08 00:00:00  182.88     154

In [10]:
df.shape

(183978, 42)

In [11]:
df_player.shape

(11060, 7)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183978 entries, 0 to 183977
Data columns (total 42 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   id                   183978 non-null  int64  
 1   player_fifa_api_id   183978 non-null  int64  
 2   player_api_id        183978 non-null  int64  
 3   date                 183978 non-null  object 
 4   overall_rating       183142 non-null  float64
 5   potential            183142 non-null  float64
 6   preferred_foot       183142 non-null  object 
 7   attacking_work_rate  180748 non-null  object 
 8   defensive_work_rate  183142 non-null  object 
 9   crossing             183142 non-null  float64
 10  finishing            183142 non-null  float64
 11  heading_accuracy     183142 non-null  float64
 12  short_passing        183142 non-null  float64
 13  volleys              181265 non-null  float64
 14  dribbling            183142 non-null  float64
 15  curve            

There is no errant data types , BUT there missing data in several columns

In [13]:
df_player.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11060 entries, 0 to 11059
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  11060 non-null  int64  
 1   player_api_id       11060 non-null  int64  
 2   player_name         11060 non-null  object 
 3   player_fifa_api_id  11060 non-null  int64  
 4   birthday            11060 non-null  object 
 5   height              11060 non-null  float64
 6   weight              11060 non-null  int64  
dtypes: float64(1), int64(4), object(2)
memory usage: 605.0+ KB


There is no errant data types and missing data

In [14]:
#look for missing data.
df.isnull().sum()

id                        0
player_fifa_api_id        0
player_api_id             0
date                      0
overall_rating          836
potential               836
preferred_foot          836
attacking_work_rate    3230
defensive_work_rate     836
crossing                836
finishing               836
heading_accuracy        836
short_passing           836
volleys                2713
dribbling               836
curve                  2713
free_kick_accuracy      836
long_passing            836
ball_control            836
acceleration            836
sprint_speed            836
agility                2713
reactions               836
balance                2713
shot_power              836
jumping                2713
stamina                 836
strength                836
long_shots              836
aggression              836
interceptions           836
positioning             836
vision                 2713
penalties               836
marking                 836
standing_tackle     

Penalties column data is important for our question
There is 836 missing values in penalties column

In [15]:
#look for duplicated data.
df.duplicated().sum()

0

There is no duplicated data

In [17]:

df.describe()

id  player_fifa_api_id  player_api_id  overall_rating  \
count  183978.00000       183978.000000  183978.000000   183142.000000   
mean    91989.50000       165671.524291  135900.617324       68.600015   
std     53110.01825        53851.094769  136927.840510        7.041139   
min         1.00000            2.000000    2625.000000       33.000000   
25%     45995.25000       155798.000000   34763.000000       64.000000   
50%     91989.50000       183488.000000   77741.000000       69.000000   
75%    137983.75000       199848.000000  191080.000000       73.000000   
max    183978.00000       234141.000000  750584.000000       94.000000   

           potential       crossing      finishing  heading_accuracy  \
count  183142.000000  183142.000000  183142.000000     183142.000000   
mean       73.460353      55.086883      49.921078         57.266023   
std         6.592271      17.242135      19.038705         16.488905   
min        39.000000       1.000000       1.000000          1.000000   
25%        69.000000      45.000000      34.000000         49.000000   
50%        74.000000      59.000000      53.000000         60.000000   
75%        78.000000      68.000000      65.000000         68.000000   
max        97.000000      95.000000      97.000000         98.000000   

       short_passing        volleys  ...         vision      penalties  \
count  183142.000000  181265.000000  ...  181265.000000  183142.000000   
mean       62.429672      49.468436  ...      57.873550      55.003986   
std        14.194068      18.256618  ...      15.144086      15.546519   
min         3.000000       1.000000  ...       1.000000       2.000000   
25%        57.000000      35.000000  ...      49.000000      45.000000   
50%        65.000000      52.000000  ...      60.000000      57.000000   
75%        72.000000      64.000000  ...      69.000000      67.000000   
max        97.000000      93.000000  ...      97.000000      96.000000   

             marking  standing_tackle  sliding_tackle      gk_diving  \
count  183142.000000    183142.000000   181265.000000  183142.000000   
mean       46.772242        50.351257       48.001462      14.704393   
std        21.227667        21.483706       21.598778      16.865467   
min         1.000000         1.000000        2.000000       1.000000   
25%        25.000000        29.000000       25.000000       7.000000   
50%        50.000000        56.000000       53.000000      10.000000   
75%        66.000000        69.000000       67.000000      13.000000   
max        96.000000        95.000000       95.000000      94.000000   

         gk_handling     gk_kicking  gk_positioning    gk_reflexes  
count  183142.000000  183142.000000   183142.000000  183142.000000  
mean       16.063612      20.998362       16.132154      16.441439  
std        15.867382      21.452980       16.099175      17.198155  
min         1.000000       1.000000        1.000000       1.000000  
25%         8.000000       8.000000        8.000000       8.000000  
50%        11.000000      12.000000       11.000000      11.000000  
75%        15.000000      15.000000       15.000000      15.000000  
max        93.000000      97.000000       96.000000      96.000000  

[8 rows x 38 columns]

Penalties column data is important for our question
Mean penalties for players is 55
Maximum penalties for players is 96

In [18]:
df['penalties']= df['penalties'].fillna(0)

In [19]:
df['penalties'].isnull().sum()

0

In [20]:

df['penalties'].astype(int)

0         48
1         48
2         48
3         47
4         47
          ..
183973    83
183974    70
183975    70
183976    53
183977    53
Name: penalties, Length: 183978, dtype: int64

In [21]:
df_combined = df.merge(df_player, left_on='player_fifa_api_id', right_on='player_fifa_api_id', how='inner')

In [16]:
# view to check merge
df_combined.head()

id_x  player_fifa_api_id  player_api_id_x                 date  \
0     1              218353           505942  2016-02-18 00:00:00   
1     2              218353           505942  2015-11-19 00:00:00   
2     3              218353           505942  2015-09-21 00:00:00   
3     4              218353           505942  2015-03-20 00:00:00   
4     5              218353           505942  2007-02-22 00:00:00   

   overall_rating  potential preferred_foot attacking_work_rate  \
0            67.0       71.0          right              medium   
1            67.0       71.0          right              medium   
2            62.0       66.0          right              medium   
3            61.0       65.0          right              medium   
4            61.0       65.0          right              medium   

  defensive_work_rate  crossing  ...  gk_handling  gk_kicking  gk_positioning  \
0              medium      49.0  ...         11.0        10.0             8.0   
1              medium      49.0  ...         11.0        10.0             8.0   
2              medium      49.0  ...         11.0        10.0             8.0   
3              medium      48.0  ...         10.0         9.0             7.0   
4              medium      48.0  ...         10.0         9.0             7.0   

   gk_reflexes  id_y  player_api_id_y         player_name  \
0          8.0     1           505942  Aaron Appindangoye   
1          8.0     1           505942  Aaron Appindangoye   
2          8.0     1           505942  Aaron Appindangoye   
3          7.0     1           505942  Aaron Appindangoye   
4          7.0     1           505942  Aaron Appindangoye   

              birthday  height  weight  
0  1992-02-29 00:00:00  182.88     187  
1  1992-02-29 00:00:00  182.88     187  
2  1992-02-29 00:00:00  182.88     187  
3  1992-02-29 00:00:00  182.88     187  
4  1992-02-29 00:00:00  182.88     187  

[5 rows x 48 columns]

we merge two tables to obtain players names from player table and pentailes from player attributes table to build relationship between them. 

### Which player has the most penalties

In [17]:
df_combined['player_name'].value_counts()

Danilo                108
Sebastian Giovinco     96
Douglas                93
Felipe                 89
Ricardo                84
                     ... 
Pires                   2
Jose Matos              2
Tiago Rosa              2
Xhevdet Gela            2
Ryan Thomson            2
Name: player_name, Length: 10848, dtype: int64

each player has several penalties in several rows

calculate all penalties for each player by sum method.

In [18]:
df_combined.groupby('player_name').sum().penalties

player_name
Aaron Appindangoye              238.0
Aaron Cresswell                1753.0
Aaron Doran                    1574.0
Aaron Galindo                   960.0
Aaron Hughes                   1324.0
                                ...  
Zsolt Low                       427.0
Zurab Khizanishvili             512.0
Zvjezdan Misimovic              709.0
de Oliveira Cleber Monteiro     478.0
dos Santos Fabio Junior         290.0
Name: penalties, Length: 10848, dtype: float64

In [19]:
answer=df_combined.groupby('player_name').sum().penalties
df_answer=pd.DataFrame(answer)
df_answer.head()

penalties
player_name                  
Aaron Appindangoye      238.0
Aaron Cresswell        1753.0
Aaron Doran            1574.0
Aaron Galindo           960.0
Aaron Hughes           1324.0

Convet the answer in the previous step to obtain data from it

In [20]:
df_answer.describe()

penalties
count  10848.000000
mean     928.608038
std      651.243223
min       20.000000
25%      422.000000
50%      779.000000
75%     1298.000000
max     6716.000000

In [21]:
df_answer['penalties'].loc[lambda s: s == df_answer['penalties'].max()]

player_name
Sebastian Giovinco    6716.0
Name: penalties, dtype: float64

Player which had the most penalties is (Sebastian Giovinco) with (6716)

In [22]:
df_answer['penalties'].loc[lambda s: s > 6000]

player_name
Danilo                6358.0
Sebastian Giovinco    6716.0
Name: penalties, dtype: float64